javalib/src/main/scala/java/util/regex/Matcher.scala

```diff
@@ -275,5 +283,358 @@ object Matcher {
+            if regex.length >= 2 && regex(0) == '\\' && regex(1).isDigit =>
+          nodeType = RegexLeaf("\\" + mapping.get(regex.substring(1).toInt).getOrElse(0))
+        case RegexLeaf(regex) =>
+        case ParentNode(children) => children.foreach(_.transformGroupNumber(mapping))
+      }
+      this
+    }
+
+    // Assigns consecutive group numbers starting from newGroupIndex to the nodes in this subtree, in a pre-order walk.
+    // Returns 1 plus the largest assigned group number.
+    def setNewGroup(index: Int): Int = {
+      newGroup = index
+      nodeType.fold(
+          r => index + 1,
+          children => children.foldLeft(index+1) {
+            case (newIndex, child) => child.setNewGroup(newIndex)
+          })
+    }
+
+    // Recursively matched groups from a match of this regular expression
+    // to the field matched of each node.
+    // If a group did not match, stringForGroupNumber can return null
+    // This is consistent with JVM regexs
+    def setMatch(stringForGroupNumber: Int => String): Unit = {
+      this.matched = stringForGroupNumber(newGroup)
+      nodeType.fold(
+          r => (),
+          children => children.foreach(_.setMatch(stringForGroupNumber))
+      )
+    }
+
+    // Propagates the start position of this node to its descendants
+    def propagateStart(): Unit = {
+      nodeType.fold(
+        regex => (),
+        children => children.foldLeft(start) {
+          case (newStart, child) => child.setStartReturnEnd(newStart)
+        }
+      )
+    }
+
+    // Propagates the end position of this node to its descendants
+    def propagateEnd(): Unit = {
+      nodeType.fold(
+        regex => (),
+        children => children.foldRight(end) {
+          case (child, newEnd) => child.setEndReturnStart(newEnd)
+        }
+      )
+    }
+
+    def setEndReturnStart(newEnd: Int): Int = {
+      start = originalGroup match {
+        case OriginallyWrapped.Absolute() => newEnd
+        case _ =>
+          if (matched == null) -1
+          else newEnd - matched.length
+      }
+      propagateStart()
+      start
+    }
+    def setStartReturnEnd(offset: Int): Int = {
+      start = if (matched == null) -1
+              else offset
+
+      originalGroup match {
+        case OriginallyWrapped.Repeater(_) => propagateEnd()
+        case _ =>                          propagateStart()
+      }
+      originalGroup match {
+        case OriginallyWrapped.Absolute() => offset
+        case _ => end
+      }
+    }
+
+    def end: Int = start + (if (matched == null) 0 else matched.length)
+
+    def buildRegex: String =
+      originalGroup.fold(groupNum => "(", (left: String, right: String) => "((?:" + left) +
+      nodeType.     fold(regex => regex,
+        children => "(?:" + children.map(_.buildRegex).mkString("") + ")") +
+      originalGroup.fold(groupNum => ")", (left: String, right: String) => ")" + right + ")")
+
+    def obtainMapping: Map[Int, Node] = {
```

Comment:
I renamed it to `getGroupNodeMap` and the associated variable to `groupNodeMap` since it is a mapping from original group numbers to their associated node. From this mapping, we can retrieve both the new group so that we can shift backreferences, and we can also retrieve the wanted mapping group number to start.